lib/querytables.js

```diff
@@ -1,44 +1,280 @@
+     * that ordering and somehow (PG internals) the subsequent calls end up ordering the table names
+     * by their `::regclass::oid` */
+
+    let metadataQuery = `WITH cdb_table_names AS (
+        SELECT format('%s.%s', quote_ident('${encodePGName(tableArray[0].schema_name)}'),
+                               quote_ident('${encodePGName(tableArray[0].table_name)}')) as id_name
+    `;
+
+    for (let i = 1; i < tableArray.length; i++) {
+        metadataQuery += `UNION ALL
+        SELECT format('%s.%s', quote_ident('${encodePGName(tableArray[i].schema_name)}'),
+                               quote_ident('${encodePGName(tableArray[i].table_name)}')) as id_name
+        `;
+    }
 
-module.exports.getAffectedTablesFromQuery = function (pg, sql, callback) {
-    const query = `SELECT * FROM CDB_QueryTables_Updated_At($cdb_query$' ${replaceTile0(sql)} '$cdb_query$)`;
+    metadataQuery += `
+        ), cdb_table_oids AS (
+            SELECT DISTINCT id_name, id_name::regclass::oid AS reloid FROM cdb_table_names
+        ), cdb_table_metadata AS (
+            SELECT
+                quote_ident(n.nspname::text) schema_name,
+                quote_ident(c.relname::text) table_name,
+                c.relkind,
+                cdb_table_oids.*,
+                (SELECT md.updated_at FROM cartodb.CDB_TableMetadata md WHERE md.tabname = reloid) AS updated_at,
+                (CASE   WHEN relkind != 'f' THEN current_database()
+                        ELSE (
+                            SELECT option_value AS dbname FROM cdb_table_oids, pg_options_to_table((
+                                SELECT fs.srvoptions
+                                FROM pg_foreign_table ft
+                                LEFT JOIN pg_foreign_server fs ON ft.ftserver = fs.oid
+                                WHERE ft.ftrelid = cdb_table_oids.reloid
+                            )) WHERE option_name='dbname')
+                         END) AS dbname
+            FROM cdb_table_oids, pg_catalog.pg_class c
+            LEFT JOIN pg_catalog.pg_namespace n ON c.relnamespace = n.oid
+            WHERE c.oid = cdb_table_oids.reloid
+        )
+        SELECT * FROM cdb_table_metadata ORDER BY reloid;`;
 
-    pg.query(query, (err, result) => {
+    const readOnly = true;
+    pg.query(metadataQuery, {}, (err, result) => {
         if (err) {
             const msg = err.message ? err.message : err;
-            return callback(new Error('could not fetch affected tables or last updated time: ' + msg));
+            return callback(new Error('Could not fetch metadata about the affected tables: ' + msg));
         }
-        result = result || {};
         const rows = result.rows || [];
+        const getForeignTableUpdatedAtPromise = promisify(getForeignTableUpdatedAt);
+
+        /* For foreign tables we need to do extra queries to extract the updated_at properly */
+        const promiseCollection = rows
```

Comment:
Avoid change objects by reference, instead, create a new array with the data augmented.
```suggestion
        const tableMetadataWithUpdatedAtPromises = rows
            .map(row => row.relkind === 'f'
                ? getForeignTableUpdatedAtPromise(pg, row).then(updated_at => Object.assign({}, row, {updated_at}))
                : Promise.resolve(row)
            );
        Promise.all(tableMetadataWithUpdatedAtPromises)
            .then(tablesMetadata => callback(null, tablesMetadata))
            .catch(err => callback(err));
```